### ISSUE TITLE
Question about Latency Results in Table 5

### ISSUE THREAD

**Issue by lky-violet** [Theia-4869/CDPruner#1]
Hello, thank you for your excellent work!

However, I have some concerns regarding the latency results presented in **Table 5**. Specifically, I find it surprising that **FastV** shows higher end-to-end latency compared to your method and **DivPrune**.

FastV only needs to compute the attention between the **last token** and **all image tokens** (i.e., target attention), which can be efficiently computed in parallel using matrix operations. After that, it only performs a single `topkmax` operation. In contrast, both your method and DivPrune involve additional computations and greedy search to find optimal solutions, which should theoretically be more time-consuming.

Based on my own experiments on **LLaVA-COCO**, here are the **end-to-end latencies (per sample)** measured under **FlashAttention**:

### LLaVA-v1.5-7B
| Method     | Latency (s) |
|------------|-------------|
| FastV      | 2.11        |
| DivPrune   | 2.33        |
| Baseline   | 2.24        |

### LLaVA-v1.6-7B
| Method     | Latency (s) |
|------------|-------------|
| FastV      | 3.77        |
| DivPrune   | 5.00        |
| Baseline   | 4.24        |

Given these results, I wonder if there might be an issue with the **baseline implementations** in your evaluation. Could you kindly clarify how the baselines (especially FastV) were implemented and how the latency was measured?

Thanks again for your great work, and I look forward to your response!

**Comment by lky-violet** [comment]
You can refer to the "Efficiency Analysis of PyramidDrop" section in PyramidDrop for adapting FlashAttention. Any attention-based method can use FlashAttention at each layer.

Also, your explanation does not account for why the latency of DivPrune is better than FastV. If the additional overhead of MMDP is excluded, the results are very similar to those reported by you.

**Comment by Theia-4869** [comment]
In Section 3.2 “Efficiency Analysis of PyramidDrop” from the PyramidDrop paper, the authors mention FlashAttention but do not explain how it is implemented with PyramidDrop. In the code provided by the authors ([pdrop_rank_drop](https://github.com/Cooperx521/PyramidDrop/blob/main/llava/model/modeling_llama_pdrop.py#L1291)), it can be observed that, in order to obtain attention scores, they re-implement eager attention in a naïve manner and repeatedly compute the attention scores, which undoubtedly adds to the inference latency.

For DivPrune, we also introduce optimizations based on the original code provided by the authors as follows:
```
image_normalized = (image_features / image_features.norm(dim=-1, keepdim=True))
cosine_matrix = 1.0 - torch.matmul(image_normalized, image_normalized.transpose(1, 2))

select_idx = torch.empty((B, self.visual_token_num), dtype=torch.long, device=device)
for i in range(self.visual_token_num):
    m2 = cosine_matrix if i==0 \
        else cosine_matrix[torch.arange(B).unsqueeze(1).expand(-1, i), select_idx[:, :i]]
    scores = torch.topk(m2, 2, dim=1, largest=False).values[:, 1] \
        if i==0 else torch.min(m2, dim=1).values
    select_idx[:, i] = torch.argmax(scores, dim=-1)

select_idx = torch.sort(select_idx).values
index_masks = torch.zeros(B, N, dtype=torch.bool, device=device)
index_masks.scatter_(1, select_idx, True)
```

With this optimization, the computational overhead of MMDP can be further reduced. You can integrate this part into our code to test the delay if interested.

Regarding the results you reported, I’m quite curious about what the “Baseline” refers to. If it denotes the unpruned LLaVA model, it seems that the latency of DivPrune after pruning is actually higher, which appears inconsistent with the expected results. As shown in Table 2 of Section 4.5 in the DivPrune paper, the inference latency of the model after pruning should be lower than that before pruning.

**Comment by lky-violet** [comment]
Thanks for your timely reply!

I'd like to point out that the extra overhead introduced by pymiddrop is virtually negligible. You might consider using matrix operations to compute the target attention of the last token — this approach is highly efficient due to PyTorch's matrix operation optimizations, and its computational complexity is still O(n).

I'm glad to see that you've optimized the `DivPrune` testing implementation.  In my own experiments, I also applied similar optimizations:

```python
def my_div_prune(self, hidden_states, k, select_indices=None):
    hidden_states_normalized = F.normalize(hidden_states, p=2, dim=1)
    # hidden_states_normalized = hidden_states.clone()
    distance_metric = torch.matmul(hidden_states_normalized, hidden_states_normalized.transpose(0, 1))
    distance_metric = -distance_metric + 1
    distance_metric.fill_diagonal_(float('inf'))

    if select_indices is None:
        select_indices = []
    else:
        distance_metric[:, select_indices] = float('-inf')

    num = 0
    while num < k:
        if len(select_indices) == 0:
            # 计算每一行的最小值
            row_min_values, row_min_indices = distance_metric.min(dim=1)
            max_value, max_index = row_min_values.max(dim=0)
            select_indices.append(max_index)
            # 把对应列设置为 -inf
            distance_metric[:, max_index] = float('-inf')
        else:
            # 简化逻辑：求集合外与集合内元素之间的 minmax
            selected_dis = distance_metric[select_indices, :]  # shape: [len(select_indices), N]
            min_values, _ = torch.min(selected_dis, dim=0)
            _, max_index = min_values.max(dim=0)
            select_indices.append(max_index)
            distance_metric[:, max_index] = float('-inf')
        num += 1

    return torch.tensor(select_indices)


def CDF_Div_prune(self, image_features, visual_token_num):
    B = image_features.shape[0]
    image_normalized = image_features / image_features.norm(dim=-1, keepdim=True)

    # 计算余弦距离矩阵（1 - cosine similarity）
    # 结果形状为 [B, N, N]
    cosine_matrix = 1.0 - torch.matmul(image_normalized, image_normalized.transpose(1, 2))

    # 初始化选择索引
    select_idx = torch.empty((B, visual_token_num), dtype=torch.long, device=image_features.device)

    # 逐步选择 token
    for i in range(visual_token_num):
        if i == 0:
            # 第一次直接用整张 cosine matrix
            m2 = cosine_matrix
            # 选择第二小的距离（因为第一个是自己）
            scores = torch.topk(m2, 2, dim=1, largest=False).values[:, 1]
        else:
            # 之后选择与已选 token 的最小余弦距离
            batch_indices = torch.arange(B).unsqueeze(1).expand(-1, i)  # shape: [B, i]
            selected = select_idx[:, :i]  # shape: [B, i]
            m2 = cosine_matrix[batch_indices, selected]  # shape: [B, i, N]
            scores = torch.min(m2, dim=1).values  # shape: [B, N]

        # 选择得分最高的 token（距离最远）
        select_idx[:, i] = torch.argmax(scores, dim=-1)

    return select_idx
```


Below is a performance comparison on the **LLaVA-COCO** dataset using **llava-v1.5-7b**:

| Method              | Inference Time per Sample |
|---------------------|---------------------------|
| `CDP_DivPrune`      | 2.43s                     |
| `My_DivPrune`       | 2.33s                     |
| `llava-v1.5-7b`     | 2.24s                     |

This shows that even with optimizations, `DivPrune` **still introduces extra latency**.

Regarding your point that `DivPrune` introduces latency — I believe the original paper refers to a **PD-separated setting** with an application focus on **video understanding**, which may explain the difference in overhead expectations.

**Comment by lky-violet** [comment]
By the way, have you noticed that running DivPrune on QA-style datasets like GQA or POPE can feel noticeably slower?

These datasets often generate only one or two tokens per sample, which means the overhead introduced by DivPrune becomes more significant relative to the total generation time.

In contrast, tasks like COCO captioning tend to produce much longer responses, which naturally amplifies the benefits of DivPrune and masks the latency cost.

This difference in generation length may partially explain why DivPrune seems more favorable on captioning benchmarks but less efficient on QA-style datasets.

**Comment by Theia-4869** [comment]
I have actually evaluated the efficiency of DivPrune on some general visual QA datasets. The MMDP process does introduce some additional overhead, but after optimization, this overhead is relatively small (less than 10ms per sample). Given this, I still think that the inference latency after pruning with DivPrune should not exceed that of the original LLaVA model.

The effiency gain from pruning may be less noticeable on smaller models (7B vs. 13B), and it might also be due to the relatively large number of visual tokens retained. When many tokens are kept, the benefits of pruning diminish, while the extra computation introduced by MMDP may outweigh the savings, potentially leading to higher latency than the unpruned model. However, under larger language models or more aggressive pruning ratios, DivPrune can still provide significant improvements in inference efficiency.

**Comment by Theia-4869** [comment]
If you find our work helpful, welcome to star our repository.
I’ll close this issue for now. If you have any further questions, feel free to reopen the issue and ask. 😊

**Comment by HitBadTrap** [comment]
Hello! Is the 10ms delay here measured on LLaVA-1.5-7B? What is the delay in video comprehension tasks, such as LLaVA-Video-7B.
